README.md
# ClassLoader

[![Build Status](https://travis-ci.org/Prowect/ClassLoader.svg)](https://travis-ci.org/Prowect/ClassLoader)
[![Code Climate](https://codeclimate.com/github/Prowect/ClassLoader/badges/gpa.svg)](https://codeclimate.com/github/Prowect/ClassLoader)
[![Test Coverage](https://codeclimate.com/github/Prowect/ClassLoader/badges/coverage.svg)](https://codeclimate.com/github/Prowect/ClassLoader/coverage)

## Beschreibung

Der ClassLoader lädt automatisch PHP-Dateien bzw. -Klassen. Hierfür erfolgt anhand des Namespaces eine Zuordnung zu einer bestimmten Datei, welche anschließend included wird.

## Installation

1. Zuerst muss die PHP-Datei, die den ClassLoader beinhaltet included werden.
2. Anschließend muss der ClassLoader angelegt (registriert) werden.

```php
<?php
use Drips\ClassLoader\ClassLoader;
require_once 'classloader.php';
$loader = new ClassLoader;
```

## Verwendung

Neben dem Laden von Klassen mithilfe der `load`-Methode können Namespaces auch manuell registriert werden, wenn diese beispielsweise in einem anderem (Unter-)Verzeichnis liegen.

```php
<?php
$loader->registerNamespace("Drips", "core/lib");
```

Somit können Dateien, die im Namespace `Drips` liegen von `core/lib/Drips` geladen werden.

Außerdem kann ein Standardverzeichnis festgelegt werden, von welchem der ClassLoader die Dateien beziehen soll.

```php
<?php
$loader->load_dir = "core/lib";
```

Dadurch werden alle Klassen die vom ClassLoader geladen werden sollen im Verzeichnis `core/lib` gesucht.


CHANGELOG
Version 0.1.1
==============================
Features
---------------
 - `load`-Methode wurde gekürzt, indem sie auf 2 weitere Methoden ausgelagert wurde. (`autoload` und `manualload`)

Bugs
---------------
 - Auf Linux-Systemen, wurde die Groß- und Kleinschreibung von Dateinamen berücksichtig, weshalb das Laden von Klassen teilweise nicht möglich war. Aus diesem Grund können von nun an ausschließlich lowercase-Dateien geladen werden.

Version 0.1
==============================
Features
---------------
 - Automatisches Laden von PHP-Klassen bzw. PHP-Dateien anhand von Namespaces
 - Registrierung von Namespaces für bestimmte Verzeichnisse
 - Festlegen eines Lade-Verzeichnisses, von dem die PHP-Klassen geladen werden sollen


tests/testcases.php
<?php

namespace tests;

use PHPUnit_Framework_Testcase;
use Drips\ClassLoader\ClassLoader;
use tests\example\Example;
use example\Other;

// load and register classloader
require_once __DIR__.'/../src/classloader.php';
$classloader = new ClassLoader;
$classloader->registerNamespace("example", __DIR__);

class TestCases extends PHPUnit_Framework_TestCase
{
    public function testCanLoadTestClass()
    {
        $this->assertEquals(TestClass::hello(), "Hello");
    }

    public function testCanLoadExample()
    {
        $this->assertTrue(Example::isTrue());
    }

    public function testCanLoadOther()
    {
        $this->assertTrue(Other::isTrue());
    }
}


tests/testclass.php
<?php

namespace tests;

class TestClass
{
    public static function hello()
    {
        return "Hello";
    }
}


src/ClassLoader.php
<?php

/**
 * Created by Prowect
 * Author: Raffael Kessler
 * Date: 13.10.2015 - 18:30
 */

namespace Drips\ClassLoader;

/**
 * Class ClassLoader
 *
 * This class allows the automatic loading of PHP files or PHP classes based on
 * their namespaces.
 */
class ClassLoader
{
    /**
     * Specifies the directory in which according to the associated PHP files to
     * be searched.
     *
     * @var string
     */
    public $load_dir;

    /**
     * Defines valid file extensions to be searched for by those fixed.
     *
     * @var array
     */
    public $extensions = array('.php', '.class.php', '.inc.php');

    /**
     * Contains the assignment of different namespaces to different (sub)
     * directories
     *
     * @var array
     */
    protected $namespaces = array();

    /**
     * Creates a new class loader instance and registers the same classloader by
     * spl_autoload_register.
     */
    public function __construct()
    {
        spl_autoload_register(function ($class) {
            $this->load($class);
        });
    }

    /**
     * Attempts to load the given class.
     *
     * @param string $class specifies the full class name, the class to load
     */
    public function load($class)
    {
        // set include path
        $include_path = $this->load_dir;
        if(!isset($this->load_dir)){
            $include_path = ".";
        }
        set_include_path($include_path.DIRECTORY_SEPARATOR);
        echo "Current include path is: $include_path\n";
        echo "Try loading: $class\n";
        // try autoloading the right php file
        spl_autoload($class, implode(',', $this->extensions));
        // if load was not successful try loading manually
        if (!class_exists($class)) {
            $parts = explode('\\', $class);
            if (count($parts) >= 2) {
                $namespace = $parts[0];
                if (array_key_exists($namespace, $this->namespaces)) {
                    $dir = $this->namespaces[$namespace];
                    foreach ($this->extensions as $extension) {
                        $path = $dir.DIRECTORY_SEPARATOR.str_replace('\\', DIRECTORY_SEPARATOR, $class).$extension;
                        if (isset($this->load_dir)) {
                            $path = $this->load_dir.DIRECTORY_SEPARATOR.$path;
                        }
                        if (file_exists($path)) {
                            require_once $path;
                            echo "Found $class with own function\n";
                            return true;
                        }
                    }
                }
            }
        } else {
            echo "Found $class with spl_autoload\n";
            return true;
        }
        echo "Class $class could not been found\n";
        return false;
    }

    /**
     * Allows you to associate different namespaces to different directories.
     *
     * @param string $namespace sets the namespace, which is located in a subdirectory.
     * @param string $directory specifies the target directory where the files of the specified namespaces are
     */
    public function registerNamespace($namespace, $directory)
    {
        $this->namespaces[$namespace] = $directory;
    }
}


src/classloader.php
<?php

/**
 * Created by Prowect
 * Author: Raffael Kessler
 * Date: 13.10.2015 - 18:30
 */

namespace Drips\ClassLoader;

/**
 * Class ClassLoader
 *
 * This class allows the automatic loading of PHP files or PHP classes based on
 * their namespaces.
 */
class ClassLoader
{
    /**
     * Specifies the directory in which according to the associated PHP files to
     * be searched.
     *
     * @var string
     */
    public $load_dir;

    /**
     * Defines valid file extensions to be searched for by those fixed.
     *
     * @var array
     */
    public $extensions = array('.php', '.class.php', '.inc.php');

    /**
     * Contains the assignment of different namespaces to different (sub)
     * directories
     *
     * @var array
     */
    protected $namespaces = array();

    /**
     * Creates a new class loader instance and registers the same classloader by
     * spl_autoload_register.
     */
    public function __construct()
    {
        spl_autoload_register(function ($class) {
            $this->load($class);
        });
    }

    /**
     * Attempts to load the given class.
     *
     * @param string $class specifies the full class name, the class to load
     *
     * @return bool returns if loading was successful
     */
    public function load($class)
    {
        // set include path
        $include_path = $this->load_dir;
        if(!isset($this->load_dir)){
            $include_path = ".";
        }
        set_include_path($include_path.DIRECTORY_SEPARATOR);
        if(!$this->autoload($class)){
            return $this->manualload($class);
        } else {
            return true;
        }
        return false;
    }

    /**
     * tries loading given class using spl_autoload function
     *
     * @param string $class specifies the full class name, the class to load
     *
     * @return bool returns if loading was successful
     */
    public function autoload($class)
    {
        spl_autoload($class, implode(",", $this->extensions));
        return class_exists($class);
    }

    /**
     * tries loading given class manually
     *
     * @param string $class specifies the full class name, the class to load
     *
     * @return bool returns if loading was successful
     */
    public function manualload($class)
    {
        $parts = explode('\\', $class);
        if (count($parts) >= 2) {
            $namespace = $parts[0];
            if (array_key_exists($namespace, $this->namespaces)) {
                $dir = $this->namespaces[$namespace];
                foreach ($this->extensions as $extension) {
                    $path = $dir.DIRECTORY_SEPARATOR.strtolower(str_replace('\\', DIRECTORY_SEPARATOR, $class)).$extension;
                    if (isset($this->load_dir)) {
                        $path = $this->load_dir.DIRECTORY_SEPARATOR.$path;
                    }
                    if (file_exists($path)) {
                        require_once $path;
                        return true;
                    }
                }
            }
        }
        return false;
    }

    /**
     * Allows you to associate different namespaces to different directories.
     *
     * @param string $namespace sets the namespace, which is located in a subdirectory.
     * @param string $directory specifies the target directory where the files of the specified namespaces are
     */
    public function registerNamespace($namespace, $directory)
    {
        $this->namespaces[$namespace] = $directory;
    }
}
